model-server/src/test/kotlin/org/modelix/model/server/V1ApiTest.kt

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (c) 2024.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.modelix.model.server
+
+import com.google.gson.JsonParser
+import io.ktor.client.request.get
+import io.ktor.client.request.post
+import io.ktor.client.request.put
+import io.ktor.client.request.setBody
+import io.ktor.client.statement.bodyAsText
+import io.ktor.http.HttpStatusCode
+import io.ktor.server.testing.ApplicationTestBuilder
+import io.ktor.server.testing.testApplication
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import org.junit.jupiter.api.Test
+import org.modelix.authorization.installAuthentication
+import org.modelix.model.InMemoryModels
+import org.modelix.model.server.handlers.KeyValueLikeModelServer
+import org.modelix.model.server.handlers.RepositoriesManager
+import org.modelix.model.server.store.InMemoryStoreClient
+import org.modelix.model.server.store.LocalModelClient
+import org.modelix.model.server.store.forGlobalRepository
+import kotlin.test.assertEquals
+import kotlin.test.assertNotEquals
+
+class V1ApiTest {
+
+    private fun runTest(block: suspend ApplicationTestBuilder.() -> Unit) = testApplication {
+        val inMemoryModels = InMemoryModels()
+        val store = InMemoryStoreClient().forGlobalRepository()
+        val localModelClient = LocalModelClient(store)
+        val repositoriesManager = RepositoriesManager(localModelClient)
+
+        application {
+            installAuthentication(unitTestMode = true)
+            installDefaultServerPlugins()
+            KeyValueLikeModelServer(repositoriesManager, store, inMemoryModels).init(this)
+        }
+
+        block()
+    }
+
+    @Test
+    fun `counter returns different ids for same key`() = runTest {
+        val url = "/counter/a"
+
+        val response1 = client.post(url)
+        val response2 = client.post(url)
+
+        assertEquals(HttpStatusCode.OK, response1.status)
+        assertEquals(HttpStatusCode.OK, response2.status)
+        assertNotEquals(response1.bodyAsText(), response2.bodyAsText())
+    }
+
+    @Test
+    fun `events are received after subscription`() = runTest {
+        val key = "dylandog"
+        val value = "a comic book"
+
+        coroutineScope {
```

Comment:
Why do these tests need an explicit scope?

Reply:
To be able to call `async`. The get request will only be answered once the put has been performed.

Reply:
That is because of the execution semantics of `runTest`?

If the intended effect of the whole code here is to achieve:

1. GET starts polling
2. PUT happens
3. GET returns

Then I am currently not sure if that really happens that way.

First, with any general type of dispatcher, the block in the `async` might start executing only after the PUT has happened. So we have no real guarantee that this test is actually testing what it assumes to test.

Second, with the test dispatcher in particular, only the `await` will actually trigger the execution of the `async` block. So we will never end up in GET being started before the PUT. And `coroutineScope` does not magically switch to a different dispatcher as far as I know.

Reply:
I have verified that it happens that way. Note that this runs in the default dispatcher since it's not the `runTest` of `kotlinx.coroutines`.
By default async is scheduled for immediate execution. However there might be a race condition, where the PUT might happen before the GET. Though I'm not sure on that, since i haven't dived deep enough into kotlin coroutines yet.

Reply:
Can we rename `runTest` to something that not conflicts with the kotlin test `runTest` method? That's part of the confusion.

To enforce the required ordering, you can use a [Semaphore](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.sync/-semaphore/)

Reply:
done